server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/core/executor/execution/model/ExecutionGroupContext.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2023 OceanBase.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.oceanbase.odc.service.connection.logicaldatabase.core.executor.execution.model;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+
+import org.springframework.util.CollectionUtils;
+
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * @Author: Lebie
+ * @Date: 2024/8/28 17:29
+ * @Description: []
+ */
+@Slf4j
+public final class ExecutionGroupContext<T, R> {
+
+    private final Collection<ExecutionGroup<T, R>> executionGroups;
+
+    private final Map<String, ExecutionUnit<T, R>> id2ExecutionUnit;
+
+    private final ExecutorService executorService;
+
+    private Map<String, ExecutionResult<R>> executionId2Result;
+
+    @Getter
+    private int completedGroupCount = 0;
+
+    public ExecutionGroupContext(Collection<ExecutionGroup<T, R>> executionGroups, ExecutorService executorService) {
+        this.executionGroups = executionGroups;
+        this.executorService = executorService;
+        this.executionId2Result = new ConcurrentHashMap<>();
+        executionGroups.stream().flatMap(group -> group.getExecutionUnits().stream())
+                .forEach(unit -> executionId2Result.put(unit.getId(), new ExecutionResult<>(ExecutionStatus.PENDING,
+                        unit.getOrder())));
+        this.id2ExecutionUnit = executionGroups.stream().flatMap(group -> group.getExecutionUnits().stream())
+                .collect(Collectors.toMap(ExecutionUnit::getId, unit -> unit));
+    }
+
+    public void execute() throws InterruptedException {
```

Comment:
generally, context may provide the executor to run some code, but should not add implement code.
My advice is 
1. let control logic or state change logic in context
2. let business logic or compute logic in related class

Reply:
Nice suggestion, I moved the compute logic into `GroupExecutionEngine` and kept status management in context.